control-plane-operator/controllers/hostedcontrolplane/cco/operator.go

```diff
@@ -0,0 +1,144 @@
+package cco
+
+import (
+	hyperv1 "github.com/openshift/hypershift/api/v1beta1"
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/kas"
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/manifests"
+	"github.com/openshift/hypershift/support/assets"
+	"github.com/openshift/hypershift/support/certs"
+	"github.com/openshift/hypershift/support/config"
+	"github.com/openshift/hypershift/support/metrics"
+	"github.com/openshift/hypershift/support/util"
+	prometheusoperatorv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/pointer"
+)
+
+var (
+	cloudCredentialOperatorMetricsService = assets.MustService(content.ReadFile, "assets/01-service.yaml")
+	cloudCredentialOperatorDeployment     = assets.MustDeployment(content.ReadFile, "assets/03-deployment.yaml")
+)
+
+func cloudCredentialOperatorLabels() map[string]string {
+	return map[string]string{"app": "cloud-credential-operator", hyperv1.ControlPlaneComponent: "cloud-credential-operator"}
+}
+
+func ReconcileCloudCredentialOperatorMetricsService(svc *corev1.Service, ownerRef config.OwnerRef) error {
+	ownerRef.ApplyTo(svc)
+
+	// The service is assigned a cluster IP when it is created.
+	// This field is immutable as shown here: https://github.com/kubernetes/api/blob/62998e98c313b2ca15b1da278aa702bdd7b84cb0/core/v1/types.go#L4114-L4130
+	// As such, to avoid an error when updating the object, only update the fields CCO specifies.
+	cloudCredentialOperatorMetricsServiceDeepCopy := cloudCredentialOperatorMetricsService.DeepCopy()
+	svc.Spec.Ports = cloudCredentialOperatorMetricsServiceDeepCopy.Spec.Ports
+	svc.Spec.Type = cloudCredentialOperatorMetricsServiceDeepCopy.Spec.Type
+	svc.Spec.Selector = cloudCredentialOperatorMetricsServiceDeepCopy.Spec.Selector
+	svc.Spec.SessionAffinity = cloudCredentialOperatorMetricsServiceDeepCopy.Spec.SessionAffinity
+
+	return nil
+}
+
+func ReconcileCloudCredentialOperatorDeployment(deployment *appsv1.Deployment, ownerRef config.OwnerRef, cloudCredentialOperatorImage, kubeRBACProxyImage, awsPodIdentityWebhookImage, releaseVersion string, dc config.DeploymentConfig, availabilityProberImage string, apiPort *int32) error {
+	ownerRef.ApplyTo(deployment)
+	deployment.Spec = cloudCredentialOperatorDeployment.DeepCopy().Spec
+	deployment.Spec.Template.Spec.Volumes = append(deployment.Spec.Template.Spec.Volumes, corev1.Volume{
```

Comment:
note: I see other YAMLs for other components have some of this stuff done inline, but I was hoping to keep the diff between the copied manifests and upstream as close to 0 as possible to help the next engineer that needed to refresh them